### PR TITLE
Update scipy_wrapper

### PR DESCRIPTION
**Pull request by SteveDiamond**
## Description
Updates the scipy_wrapper as suggested in https://github.com/scipy/scipy/issues/18749
Issue link (if applicable): #2158 

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [x] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [ ] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [ ] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

**Comment by SteveDiamond** [comment]
Please include #2160.

File: cvxpy/interface/scipy_wrapper.py [hunk 1]
```diff
@@ -14,10 +14,18 @@
...
 
 from cvxpy.expressions import expression as exp
 
+SPARSE_MATRIX_CLASSES = [
+    sparse.csc_matrix, 
+    sparse.csr_matrix, 
+    sparse.coo_matrix, 
+    sparse.bsr_matrix, 
+    sparse.lil_matrix, 
+    sparse.dia_matrix
```
Comment (ivirshup) [inline cvxpy/interface/scipy_wrapper.py:27]:
```suggestion
    sparse.dia_matrix,
    sparse.dok_matrix,
```

I forgot `dok` in my comment

**Review by phschiele — APPROVED** [review]
LGTM. Let me summarize the understanding that I've gathered as of now (this might be obvious to some readers).

1. Returning `NotImplemented` from a binary operation (instead of raising `NotImplementedError`) is a signal that a certain operation is not implemented, but a fallback may be available (e.g., if `a.__eq__(b)` is not implemented, `b.__eq__(a)` might still be available). Only if all recovery attempts fail, an exception is being raised.
2. `scipy_wrapper` ensures that all binary operations fall back to the cvxpy implementation when `other` is a cvxpy expression.
3. This might not always be necessary, though, as SciPy may already provide this signal in their implementation, e.g., [here](https://github.com/scipy/scipy/blob/8e55aea98ee0a89e1404acec410aeb879aef33a7/scipy/sparse/_base.py#L678).

**Comment by PTNobel** [comment]
> 1. Returning NotImplemented from a binary operation (instead of raising NotImplementedError) is a signal that a certain operation is not implemented, but a fallback may be available (e.g., if a.__eq__(b) is not implemented, b.__eq__(a) might still be available). Only if all recovery attempts fail, an exception is being raised.

Correct.

> 3.    This might not always be necessary, though, as SciPy may already provide this signal in their implementation, e.g., [here](https://github.com/scipy/scipy/blob/8e55aea98ee0a89e1404acec410aeb879aef33a7/scipy/sparse/_base.py#L678).

Do we know how recent that code was? I wonder if we even need this hotfix nowadays...It was added in 2013 per git blame, and it seems this originated as: https://github.com/cvxpy/cvxpy/commit/f8b8e7ed05bfd57b11f3234bbc1ccbf8d87b56bb#diff-5b5c7742b5ff35de17876c5a718a7f6f1aeb774037c2e73d3c39bb7e937607c3
which doesn't seem relevant to this file...

In any case, I wonder if there is still a need. Especially on the `__div__, __add__, __sub__, __mul__` arguments.

Testing with
```python
class C:
    pass
    
getattr(sp.dok_matrix(np.zeros((5, 5))), func)(C())
```

only errored for `__ge__`, `__le__`, `__lt__`, `__gt__` and returned `NotImplemented` for the rest.

**Comment by phschiele** [comment]
> > 1. Returning NotImplemented from a binary operation (instead of raising NotImplementedError) is a signal that a certain operation is not implemented, but a fallback may be available (e.g., if a.**eq**(b) is not implemented, b.**eq**(a) might still be available). Only if all recovery attempts fail, an exception is being raised.
> 
> Correct.
> 
> > 3. This might not always be necessary, though, as SciPy may already provide this signal in their implementation, e.g., [here](https://github.com/scipy/scipy/blob/8e55aea98ee0a89e1404acec410aeb879aef33a7/scipy/sparse/_base.py#L678).
> 
> Do we know how recent that code was? I wonder if we even need this hotfix nowadays...It was added in 2013 per git blame, and it seems this originated as: [f8b8e7e#diff-5b5c7742b5ff35de17876c5a718a7f6f1aeb774037c2e73d3c39bb7e937607c3](https://github.com/cvxpy/cvxpy/commit/f8b8e7ed05bfd57b11f3234bbc1ccbf8d87b56bb#diff-5b5c7742b5ff35de17876c5a718a7f6f1aeb774037c2e73d3c39bb7e937607c3) which doesn't seem relevant to this file...
> 
> In any case, I wonder if there is still a need. Especially on the `__div__, __add__, __sub__, __mul__` arguments.
> 
> Testing with
> 
> ```python
> class C:
>     pass
>     
> getattr(sp.dok_matrix(np.zeros((5, 5))), func)(C())
> ```
> 
> only errored for `__ge__`, `__le__`, `__lt__`, `__gt__` and returned `NotImplemented` for the rest.

Good point. Indeed, I have checked that, based on our tests, we do not need any patch with recent SciPy versions. However, because I neither know that our tests cover all cases, nor checked with older SciPy versions, I was hesitant to suggest removing the patching entirely.

**Comment by PTNobel** [comment]
> However, because I neither know that our tests cover all cases, nor checked with older SciPy versions, I was hesitant to suggest removing the patching entirely.

Could we deprecate the patch in 1.4 and see if it is still needed? We could raise a warning and asks people to contact us if they hit it.

**Comment by rileyjmurray** [comment]
@PTNobel I'd support deprecating in 1.4.

**Pull request merged**